### PR TITLE
Stop Facebook contact if we fail 10 consecutives times

### PR DESCRIPTION
**Pull request by norkans7**
Proposed solution to the Facebook users who block pages to send them messages

We stop the contact after we fail 10 consecutive times

### CODE REVIEW

File: temba/channels/types/facebook/__init__.py [hunk 1]
```diff
@@ -97,6 +97,15 @@ def send(self, channel, msg, text):
                 raise SendException(six.text_type(e), event=event, start=start)
 
         if response.status_code != 200:
+            contact_obj = Contact.objects.get(id=msg.contact)
+            last_ten_messages = contact_obj.msgs.all().exclude(pk=msg.id)[:10].values('direction', 'status')
```
Comment (nicpottier) [inline temba/channels/types/facebook/__init__.py]: We should make sure this hits a perfect index, time bounding might also make sense here.. IE, 10 failures in the last month?

File: temba/channels/types/facebook/__init__.py [hunk 2]
```diff
@@ -97,6 +97,15 @@ def send(self, channel, msg, text):
...
+            recent_success = len(last_ten_messages) < 10
+            for last_msg in last_ten_messages:
+                recent_success = recent_success or last_msg['direction'] == INCOMING or last_msg['status'] not in [FAILED, ERRORED]
+
+            if not recent_success:
+                contact_obj.stop(contact_obj.created_by)
```
Comment (nicpottier) [inline temba/channels/types/facebook/__init__.py]: Sure seems like some fancy SQL should be able to do this no? Can we do a limit then a count of each status? That might be slightly quicker. At the very least we don't need to do any iteration here if the length of the list is less than 10.

File: temba/channels/types/facebook/__init__.py (hunk 1 above)
Comment (norkans7) [inline temba/channels/types/facebook/__init__.py]: I am not considering in the last month, I thought the logic as to check the we tried to send to the contact 10 times and failed and never received a messages from them the whole time we sent tried to send them, I think if they sent a message probably they are interested and it is different issue not them blocking the page.

I'll check the index

File: temba/channels/types/facebook/__init__.py (hunk 2 above)
Comment (norkans7) [inline temba/channels/types/facebook/__init__.py]: after the limit we cannot use distinct and filter

**Comment by JPow** [comment]
@norkans7 I've written to Facebook for clarity on the error code reporting. We will revert with some guidance on the number fails before a stop if FB no help. 10 feels like a lot.

File: temba/channels/types/facebook/__init__.py [hunk 3]
```diff
@@ -97,6 +97,18 @@ def send(self, channel, msg, text):
                 raise SendException(six.text_type(e), event=event, start=start)
 
         if response.status_code != 200:
+            contact_obj = Contact.objects.get(id=msg.contact)
```
Comment (nicpottier) [inline temba/channels/types/facebook/__init__.py]: We should make this a bit more specific to the error message Facebook returns in this case. (even though that is broad) We don't want to go failing every contact if say, a facebook token expires and we receive 401s or whatnot.

**Comment by nicpottier** [comment]
@jpow we definitely want to be conservative about something like this, so I think 10 isn't a bad starting point. Failing contacts is a big action on RapidPro (removes them from all groups etc..) so we really want to make sure we can't reach them and because of the ambiguity of this error message from Facebook we want to be triple sure about that.

**Comment by norkans7** [comment]
@JPow any more clarifications you got? we have reduced the number to be 5 times

**Comment by rowanseymour** [comment]
Closing this for now due to lack of clarity about how to proceed

**Pull request closed without merging**